{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# CCM Benchmate Apis module tutorial\n",
    "\n",
    "In this notbook we will explore some of the api endpoints that are exposed in the module, how to use them and how to integrate their outputs.\n",
    "\n",
    "This is not a complete example of all the capabilities of this module, its main strength is not that you can search these enpoints in an interactive fasion and explore data as you go but that you can autommate this process in an arbitrary depth provided that you have an understanding of all the outputs that are returned and what to do when something doesn't go according to plan.\n",
    "\n",
    "## Ensembl API:\n",
    "\n",
    "This module **does not** expose all the available endpoints. There are alot of them and not all of them seemed to be immediately useful. If this is not the case please create an issue and we'll look into it."
   ],
   "id": "9f203506c68ee025"
  },
  {
   "metadata": {
    "collapsed": true
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "\n",
    "# load the module\n",
    "\n",
    "from ccm_benchmate.apis.ensembl import Ensembl\n",
    "ensembl=Ensembl()"
   ],
   "id": "initial_id",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "By default the dataset is set to human, you can change this using the dataset parameter. I will continue wih human. The API uses the latest dataset and other datasets are not available. Please keep this in mind if your queries are not identical at different time points. It is possible to install older versions locally.\n",
    "\n",
    "If this is something you are interested in please contact CCM to discuss next steps.\n",
    "\n",
    "### Variation\n",
    "\n",
    "This enpoint is actually 3 endpoints joined together, you have 3 options, you can search for variant using their variant id, additionally, you can search by pubmed or pmc id (if you know the id of the paper) and it will return all the variants that are associated with that paper."
   ],
   "id": "30ded3e75d12a136"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "info = ensembl.variation(\"rs56116432\", add_annotations=True)\n",
    "info"
   ],
   "id": "16e9b97e3cef398c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The add_annotation feature returns a lot of information about the variant. If you do not use that your results will be much smaller.",
   "id": "c5b459676fa0c32a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "info = ensembl.variation(\"rs56116432\")\n",
    "info"
   ],
   "id": "3805b4235cc311bb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "info_pub = ensembl.variation(\"26318936\", method=\"publication\", pubtype=\"pubmed\")\n",
    "info_pub"
   ],
   "id": "efd7ecd2c4a5605d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "This publication method returns all the ids mentioned in a paper, you can then use these ids and search for variant information if you'd like. Just be nice and do not overwhelm the ensembl api endpoint. It's usually a good idea to wait a second between a few queries. If you are doing this as part of a script this can be done easily using `sys.sleep`.\n",
    "\n",
    "The translate enpoint translates the variantn to different annotaions."
   ],
   "id": "ae5d9aa82da2fe61"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "notations=ensembl.variation(\"rs56116432\", method='translate')\n",
    "notations"
   ],
   "id": "b91cf722cd7ebfad",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The vep module runs VEP, that's it. There are a lot of tools available for you to choose from for annotations. The only downside is your variant needs to be a `ccm_benchmate.variant.Variant` instance. This is not that big of a deal as we will see.",
   "id": "dec1093384226993"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from ccm_benchmate.variant.variant import SequenceVariant\n",
    "# See variant module for more information\n",
    "myvar= SequenceVariant(1, 55051215, 'G', 'GA')\n",
    "\n",
    "vep_info = ensembl.vep(species=\"human\", variant=myvar, tools=None)\n",
    "vep_info"
   ],
   "id": "419de259f3ed03e0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Here is a list of all the available tools, not all of them can be used all at once, some of them are mutually exclusive. If a tool doesnt apply to your variant (like you have an intergenic variant you will not get an alphamissense score), that filed will not be in the returned `JSON` object. Do not rely on existence of keys. This is true for all the apis\n",
    "incorporated in this module, that is just how the apis work in these endpoints.\n",
    "\n",
    "The phenotype module returns information about phenotypes that are associated with a genomic range. For that you will\n",
    "need a `GenomicRanges` instance, please see its readme and notebook for more information."
   ],
   "id": "c6345770bb4f3440"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from ccm_benchmate.ranges.genomicranges import GenomicRange\n",
    "grange = GenomicRange(9, 22125503, 22125520, \"+\")\n",
    "phenotypes = ensembl.phenotype(grange)\n",
    "phenotypes"
   ],
   "id": "61f73281821d45e0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "pd.DataFrame(phenotypes)"
   ],
   "id": "9694bc07f127c78",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "As you can see there is a lot of information here, we can look at one of them to see what's in it.m",
   "id": "4a5ea4ec1339ab44"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "phenotypes[1]",
   "id": "c6fe36a8513113fe",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "We can see all the information that comes with it, there are some common ones ilke location, and ontology accessions, additionally you can see if there are papers associated with this annotation. The locations, can then be further used to search for what is in that region to see what other genomic features are present, the papers can be used in the literature module to ingest and create a knowledgebase (see its notebook and documentation) and the ids can be used in the ncbi api to get more information about them.\n",
    "\n",
    "There  are couple of straightforward enpoints we will just go through them quickly"
   ],
   "id": "4eda3bad73a6102b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "seq = ensembl.sequence(\"ENSG00000139618\")\n",
    "seq"
   ],
   "id": "c3cc0e7b73090576",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from ccm_benchmate.ranges.genomicranges import GenomicRange\n",
    "grange = GenomicRange(9, 22125503, 22125520, \"+\")\n",
    "\n",
    "overlap = ensembl.overlap(grange, features=[\"transcript\"])\n",
    "overlap"
   ],
   "id": "de75f444fd9b5ee0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "There are a lot of features available to search from but not all of them are availble at any given time for a given region. Sometimes you want to convert some coordinate system to antother, like genomic coordinates to protein coordinates, this can be annoying to code your own, that's why we have the mapping endpoint.",
   "id": "157ae2d4c8f93c54"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "ensembl.mapping(\"ENST00000650946\", 100, 120, type=\"cDNA\")",
   "id": "7eb21af96cbde03c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "The last one we are going explore I think is the most useful one. The `xrefs` endpoint returns all the other databases that the ensembl api cross references to. You can then take these ids and use the api enpoints that are presented here or in other places to deepend your research.",
   "id": "bc22cc03f7611658"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "xrefs = ensembl.xrefs(\"ENSG00000139618\")\n",
    "xrefs"
   ],
   "id": "67584a4a3fa2e7b1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Finally, you can return about the species, and the kinds of infromation that is available in the api (there may be changes and that is beyond our control) using `Ensembl.info` method.\n",
    "\n",
    "## Stringdb API\n",
    "\n",
    "Stringdb is a web platform that focuses on protein-protein interactions, you will need to specify your species and protein identifiers. I've also built in an option to run the interaction queires recursively. That is, you can take a protein and gather all the other proteins that interact with it, then take them all and repeat the process to generate a network of arbitrary depth. Of course this will increase the number things returned exponentially and will talke exponentially longer. So keep that in mind."
   ],
   "id": "3a16dafca9521b2b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from ccm_benchmate.apis.stringdb import StringDb\n",
    "\n",
    "stringdb=StringDb()"
   ],
   "id": "5052c13e0b63b05c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "network = stringdb.gather(\"human\", name=\"ENSP00000354587\", get_network=False)\n",
    "pd.DataFrame(network)"
   ],
   "id": "47fa1168f0d1c651",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Get network specifies whether you wanto get the itneractors of interactor. If you specify that to `True` and network depth. The number will grow exponentially, so anything over 3 is probably overkill by a wide margin. You can use a wide range of identifiers, in the example above we are using an ensembl protein id (things need to be proteins) but it can be a whole bunch of other ids. See their documentation for details.\n",
    "\n",
    "Speaking of protien-protein interactions we have 2 more apis implemented.\n",
    "\n",
    "## BioGrid\n",
    "\n",
    "Biogrid is a similar platform that focuses on protein-proteininteractions with some experimental data annotaions as to how that interaction is determined. To use biogrid you need to get an access key but it is free."
   ],
   "id": "bf64e0d385b17373"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-18T17:37:31.892867Z",
     "start_time": "2025-06-18T17:37:24.775178Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from ccm_benchmate.apis.others import BioGrid\n",
    "biogrid=BioGrid(access_key=\"<your api key>\")"
   ],
   "id": "e6ec330b621fe2af",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Inital configuration collects a bunch of up to date information so it might take a few seconds. This information includes\n",
    "the kinds of evidences that are collectes, organisms and types of indentifiers that are supported. This is useful because we will need to specify these things when we query the database."
   ],
   "id": "c93f49efb37034ba"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-18T17:38:18.807900Z",
     "start_time": "2025-06-18T17:38:17.537241Z"
    }
   },
   "cell_type": "code",
   "source": [
    "interactions=biogrid.interactions(gene_list=[\"ENSP00000354587\"])\n",
    "interactions"
   ],
   "id": "a5288cedb7c05b8d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      BIOGRID_INTERACTION_ID ENTREZ_GENE_A ENTREZ_GENE_B  BIOGRID_ID_A  \\\n",
       "0                        103          6416          2318        112315   \n",
       "1                        117         84665            88        124185   \n",
       "2                        183            90          2339        106605   \n",
       "3                        278          2624          5371        108894   \n",
       "4                        418          6118          6774        112038   \n",
       "...                      ...           ...           ...           ...   \n",
       "9995                   40341         36328         39088         62105   \n",
       "9996                   40342         36328       3354918         62105   \n",
       "9997                   40343         36327         33616         62104   \n",
       "9998                   40344         36327         43665         62104   \n",
       "9999                   40345         36327         50379         62104   \n",
       "\n",
       "      BIOGRID_ID_B SYSTEMATIC_NAME_A SYSTEMATIC_NAME_B OFFICIAL_SYMBOL_A  \\\n",
       "0           108607                 -                 -            MAP2K4   \n",
       "1           106603                 -                 -              MYPN   \n",
       "2           108625                 -                 -             ACVR1   \n",
       "3           111384                 -                 -             GATA2   \n",
       "4           112651       RP4-547C9.3                 -              RPA2   \n",
       "...            ...               ...               ...               ...   \n",
       "9995         64481      Dmel_CG13167       Dmel_CG3922           Vha36-2   \n",
       "9996         78043      Dmel_CG13167      Dmel_CG17420           Vha36-2   \n",
       "9997         59817      Dmel_CG13168      Dmel_CG15423           CG13168   \n",
       "9998         68515      Dmel_CG13168      Dmel_CG15549           CG13168   \n",
       "9999         72662      Dmel_CG13168      Dmel_CG15578           CG13168   \n",
       "\n",
       "     OFFICIAL_SYMBOL_B                                         SYNONYMS_A  \\\n",
       "0                 FLNC  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...   \n",
       "1                ACTN2                             CMD1DD|CMH22|MYOP|RCM4   \n",
       "2                 FNTA            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI   \n",
       "3                  PML                           DCML|IMD21|MONOMAC|NFE1B   \n",
       "4                STAT3                      REPA2|RP-A p32|RP-A p34|RPA32   \n",
       "...                ...                                                ...   \n",
       "9995             RpS17                               CG13167|Dmel\\CG13167   \n",
       "9996             RpL15                               CG13167|Dmel\\CG13167   \n",
       "9997           CG15423                                       Dmel\\CG13168   \n",
       "9998           CG15549                                       Dmel\\CG13168   \n",
       "9999           CG15578                                       Dmel\\CG13168   \n",
       "\n",
       "      ... PUBMED_ID ORGANISM_A ORGANISM_B       THROUGHPUT  QUANTITATION  \\\n",
       "0     ...   9006895       9606       9606   Low Throughput             -   \n",
       "1     ...  11309420       9606       9606   Low Throughput             -   \n",
       "2     ...   8599089       9606       9606   Low Throughput             -   \n",
       "3     ...  10938104       9606       9606   Low Throughput             -   \n",
       "4     ...  10875894       9606       9606   Low Throughput             -   \n",
       "...   ...       ...        ...        ...              ...           ...   \n",
       "9995  ...  14605208       7227       7227  High Throughput             -   \n",
       "9996  ...  14605208       7227       7227  High Throughput             -   \n",
       "9997  ...  14605208       7227       7227  High Throughput             -   \n",
       "9998  ...  14605208       7227       7227  High Throughput             -   \n",
       "9999  ...  14605208       7227       7227  High Throughput             -   \n",
       "\n",
       "      MODIFICATION  ONTOLOGY_TERMS QUALIFICATIONS TAGS SOURCEDB  \n",
       "0                -              {}              -    -  BIOGRID  \n",
       "1                -              {}              -    -  BIOGRID  \n",
       "2                -              {}              -    -  BIOGRID  \n",
       "3                -              {}              -    -  BIOGRID  \n",
       "4                -              {}              -    -  BIOGRID  \n",
       "...            ...             ...            ...  ...      ...  \n",
       "9995             -              {}              -    -  BIOGRID  \n",
       "9996             -              {}              -    -  BIOGRID  \n",
       "9997             -              {}              -    -  BIOGRID  \n",
       "9998             -              {}              -    -  BIOGRID  \n",
       "9999             -              {}              -    -  BIOGRID  \n",
       "\n",
       "[10000 rows x 24 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOGRID_INTERACTION_ID</th>\n",
       "      <th>ENTREZ_GENE_A</th>\n",
       "      <th>ENTREZ_GENE_B</th>\n",
       "      <th>BIOGRID_ID_A</th>\n",
       "      <th>BIOGRID_ID_B</th>\n",
       "      <th>SYSTEMATIC_NAME_A</th>\n",
       "      <th>SYSTEMATIC_NAME_B</th>\n",
       "      <th>OFFICIAL_SYMBOL_A</th>\n",
       "      <th>OFFICIAL_SYMBOL_B</th>\n",
       "      <th>SYNONYMS_A</th>\n",
       "      <th>...</th>\n",
       "      <th>PUBMED_ID</th>\n",
       "      <th>ORGANISM_A</th>\n",
       "      <th>ORGANISM_B</th>\n",
       "      <th>THROUGHPUT</th>\n",
       "      <th>QUANTITATION</th>\n",
       "      <th>MODIFICATION</th>\n",
       "      <th>ONTOLOGY_TERMS</th>\n",
       "      <th>QUALIFICATIONS</th>\n",
       "      <th>TAGS</th>\n",
       "      <th>SOURCEDB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>...</td>\n",
       "      <td>11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>...</td>\n",
       "      <td>8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>...</td>\n",
       "      <td>10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>...</td>\n",
       "      <td>10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>40341</td>\n",
       "      <td>36328</td>\n",
       "      <td>39088</td>\n",
       "      <td>62105</td>\n",
       "      <td>64481</td>\n",
       "      <td>Dmel_CG13167</td>\n",
       "      <td>Dmel_CG3922</td>\n",
       "      <td>Vha36-2</td>\n",
       "      <td>RpS17</td>\n",
       "      <td>CG13167|Dmel\\CG13167</td>\n",
       "      <td>...</td>\n",
       "      <td>14605208</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>40342</td>\n",
       "      <td>36328</td>\n",
       "      <td>3354918</td>\n",
       "      <td>62105</td>\n",
       "      <td>78043</td>\n",
       "      <td>Dmel_CG13167</td>\n",
       "      <td>Dmel_CG17420</td>\n",
       "      <td>Vha36-2</td>\n",
       "      <td>RpL15</td>\n",
       "      <td>CG13167|Dmel\\CG13167</td>\n",
       "      <td>...</td>\n",
       "      <td>14605208</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>40343</td>\n",
       "      <td>36327</td>\n",
       "      <td>33616</td>\n",
       "      <td>62104</td>\n",
       "      <td>59817</td>\n",
       "      <td>Dmel_CG13168</td>\n",
       "      <td>Dmel_CG15423</td>\n",
       "      <td>CG13168</td>\n",
       "      <td>CG15423</td>\n",
       "      <td>Dmel\\CG13168</td>\n",
       "      <td>...</td>\n",
       "      <td>14605208</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>40344</td>\n",
       "      <td>36327</td>\n",
       "      <td>43665</td>\n",
       "      <td>62104</td>\n",
       "      <td>68515</td>\n",
       "      <td>Dmel_CG13168</td>\n",
       "      <td>Dmel_CG15549</td>\n",
       "      <td>CG13168</td>\n",
       "      <td>CG15549</td>\n",
       "      <td>Dmel\\CG13168</td>\n",
       "      <td>...</td>\n",
       "      <td>14605208</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>40345</td>\n",
       "      <td>36327</td>\n",
       "      <td>50379</td>\n",
       "      <td>62104</td>\n",
       "      <td>72662</td>\n",
       "      <td>Dmel_CG13168</td>\n",
       "      <td>Dmel_CG15578</td>\n",
       "      <td>CG13168</td>\n",
       "      <td>CG15578</td>\n",
       "      <td>Dmel\\CG13168</td>\n",
       "      <td>...</td>\n",
       "      <td>14605208</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>High Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 24 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "biogrid.organisms",
   "id": "191a0fe8d0195f3",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## IntAct\n",
    "\n",
    "Intact is one other interaction database. There were a lot of requests to include all of these in the package, While they provide similar information they do have different use cases."
   ],
   "id": "1fa01c39b9dca4c8"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-18T18:59:08.081183Z",
     "start_time": "2025-06-18T18:59:07.787779Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from ccm_benchmate.apis.others import IntAct\n",
    "intact=IntAct(page_size=100)"
   ],
   "id": "613bcd08950b03c7",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-06-18T18:59:58.262585Z",
     "start_time": "2025-06-18T18:59:39.079811Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# to search intact you need the ebi id, this you can get from ensembl.xrefs or from uniprot (see part2)\n",
    "interactions=intact.intact_search(\"Q05471\")\n",
    "interactions"
   ],
   "id": "ec79c7e9a7e5d1ba",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                    idA                 idB  taxidA  taxidB  \\\n",
       "0    Q05471 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "1    Q05471 (uniprotkb)  Q06707 (uniprotkb)  559292  559292   \n",
       "2    Q05471 (uniprotkb)  Q06707 (uniprotkb)  559292  559292   \n",
       "3    P38326 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "4    P31376 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "..                  ...                 ...     ...     ...   \n",
       "243  Q03940 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "244  Q05471 (uniprotkb)  Q03940 (uniprotkb)  559292  559292   \n",
       "245  Q03940 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "246  Q03940 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "247  Q12464 (uniprotkb)  Q05471 (uniprotkb)  559292  559292   \n",
       "\n",
       "    experimental_role_A experimental_role_B stoichiometry_A stoichiometry_B  \\\n",
       "0                  bait                prey             0-0             0-0   \n",
       "1                  bait                prey             0-0             0-0   \n",
       "2                  bait                prey             0-0             0-0   \n",
       "3                  bait                prey             0-0             0-0   \n",
       "4                  bait                prey             0-0             0-0   \n",
       "..                  ...                 ...             ...             ...   \n",
       "243                bait                prey             1-1             1-1   \n",
       "244                prey                bait             1-1             1-1   \n",
       "245                bait                prey             0-0             0-0   \n",
       "246                prey                bait             1-1             1-1   \n",
       "247                prey                bait             1-1             1-1   \n",
       "\n",
       "    detection_method                                        annotations  \\\n",
       "0                tap  author-list (Lambert J-P, Fillingham J. , Gree...   \n",
       "1                tap  curation depth (IMEx)\\npublication year (2004)...   \n",
       "2      anti tag coip  accepted (Accepted 2024-MAY-08 AT 09:40 BST AT...   \n",
       "3      anti tag coip  accepted (Accepted 2024-MAY-08 AT 09:40 BST AT...   \n",
       "4      anti tag coip  accepted (Accepted 2024-MAY-08 AT 09:40 BST AT...   \n",
       "..               ...                                                ...   \n",
       "243              tap  author-confidence (z-score: 0.683760282208367)...   \n",
       "244              tap  author-confidence (probability: 99.6)\\nuniprot...   \n",
       "245              tap  accepted (null)\\nfigure legend (Supplemental T...   \n",
       "246              tap  author-confidence (probability: 99.4)\\nuniprot...   \n",
       "247              tap  number of peptides (8)\\nauthor-confidence (pro...   \n",
       "\n",
       "     is_negative  affected_by_mutation pubmed_id  score  \n",
       "0          False                 False  21179020   0.35  \n",
       "1          False                 False  15045029   0.80  \n",
       "2          False                 False  37968396   0.80  \n",
       "3          False                 False  37968396   0.85  \n",
       "4          False                 False  37968396   0.85  \n",
       "..           ...                   ...       ...    ...  \n",
       "243        False                 False  16554755   0.93  \n",
       "244        False                 False  16554755   0.93  \n",
       "245        False                 False  16429126   0.93  \n",
       "246        False                 False  16554755   0.93  \n",
       "247        False                 False  16554755   0.85  \n",
       "\n",
       "[248 rows x 14 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idA</th>\n",
       "      <th>idB</th>\n",
       "      <th>taxidA</th>\n",
       "      <th>taxidB</th>\n",
       "      <th>experimental_role_A</th>\n",
       "      <th>experimental_role_B</th>\n",
       "      <th>stoichiometry_A</th>\n",
       "      <th>stoichiometry_B</th>\n",
       "      <th>detection_method</th>\n",
       "      <th>annotations</th>\n",
       "      <th>is_negative</th>\n",
       "      <th>affected_by_mutation</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>0-0</td>\n",
       "      <td>0-0</td>\n",
       "      <td>tap</td>\n",
       "      <td>author-list (Lambert J-P, Fillingham J. , Gree...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>21179020</td>\n",
       "      <td>0.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>Q06707 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>0-0</td>\n",
       "      <td>0-0</td>\n",
       "      <td>tap</td>\n",
       "      <td>curation depth (IMEx)\\npublication year (2004)...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>15045029</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>Q06707 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>0-0</td>\n",
       "      <td>0-0</td>\n",
       "      <td>anti tag coip</td>\n",
       "      <td>accepted (Accepted 2024-MAY-08 AT 09:40 BST AT...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>37968396</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P38326 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>0-0</td>\n",
       "      <td>0-0</td>\n",
       "      <td>anti tag coip</td>\n",
       "      <td>accepted (Accepted 2024-MAY-08 AT 09:40 BST AT...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>37968396</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P31376 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>0-0</td>\n",
       "      <td>0-0</td>\n",
       "      <td>anti tag coip</td>\n",
       "      <td>accepted (Accepted 2024-MAY-08 AT 09:40 BST AT...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>37968396</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>Q03940 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>1-1</td>\n",
       "      <td>1-1</td>\n",
       "      <td>tap</td>\n",
       "      <td>author-confidence (z-score: 0.683760282208367)...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>16554755</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>Q03940 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>prey</td>\n",
       "      <td>bait</td>\n",
       "      <td>1-1</td>\n",
       "      <td>1-1</td>\n",
       "      <td>tap</td>\n",
       "      <td>author-confidence (probability: 99.6)\\nuniprot...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>16554755</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Q03940 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>bait</td>\n",
       "      <td>prey</td>\n",
       "      <td>0-0</td>\n",
       "      <td>0-0</td>\n",
       "      <td>tap</td>\n",
       "      <td>accepted (null)\\nfigure legend (Supplemental T...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>16429126</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>Q03940 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>prey</td>\n",
       "      <td>bait</td>\n",
       "      <td>1-1</td>\n",
       "      <td>1-1</td>\n",
       "      <td>tap</td>\n",
       "      <td>author-confidence (probability: 99.4)\\nuniprot...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>16554755</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Q12464 (uniprotkb)</td>\n",
       "      <td>Q05471 (uniprotkb)</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "      <td>prey</td>\n",
       "      <td>bait</td>\n",
       "      <td>1-1</td>\n",
       "      <td>1-1</td>\n",
       "      <td>tap</td>\n",
       "      <td>number of peptides (8)\\nauthor-confidence (pro...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>16554755</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>248 rows × 14 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Intact database contains information not just about protein-protein interactions but also other molecule types. This means your response could be quite large. Also I have integrated so that the api keeps searching for interactions until the last page is reached. This means you will get all the results once the request is comple but if your request has a lot of information it might take a few seconds to a few minutes.",
   "id": "1851a512500155c9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "dd87cdb3d51b780"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
